File: apps/worker/src/db/story_items_repo.ts
/**
 * Attach an item to a story.
 *
 * INSERT OR IGNORE ensures idempotency: re-runs that see already-clustered
 * items produce no side-effects.
 *
 * Returns true if the item was actually attached (i.e., it was new to this story).
 */
export async function attachItem(
  db: D1Database,
  storyId: string,
  itemId: string,
  addedAt: string,
): Promise<boolean> {
  const result = await db
    .prepare(
      `INSERT OR IGNORE INTO story_items (story_id, item_id, added_at, rank)
       VALUES (?, ?, ?, 0)`,
    )
    .bind(storyId, itemId, addedAt)
    .run();

  return (result.meta?.changes ?? 0) > 0;
}

File: apps/local-engine/db/repos/story_items_repo.py
"""db/repos/story_items_repo.py — Attach items to stories (INSERT OR IGNORE)."""

from __future__ import annotations

import aiosqlite


async def attach_item(
    db: aiosqlite.Connection,
    story_id: str,
    item_id: str,
    added_at: str,
) -> bool:
    """INSERT OR IGNORE an item into a story.

    Idempotent: re-running with already-attached items has no side-effects.

    Returns:
        True if the item was actually inserted (new attachment);
        False if it was already present (IGNORE fired).
    """
    async with db.execute(
        """
        INSERT OR IGNORE INTO story_items (story_id, item_id, added_at, rank)
        VALUES (?, ?, ?, 0)
        """,
        (story_id, item_id, added_at),
    ) as cursor:
        return cursor.rowcount > 0
